File: LeetCode/Hard/0895-maximum-frequency-stack/0895-maximum-frequency-stack.cpp
class FreqStack {
public:
    FreqStack()
    {
    }
    
    void push(int val) {
        // get val's freq - map default value is 0
        size_t val_freq = ++m_freq_map[val];

        // push to the right freq stack
        m_stack[val_freq].push(val);

        // update stack's max freq
        m_freq = std::max(m_freq, val_freq);
    }
    
    int pop() {
        // extract top value
        int top_val = m_stack[m_freq].top();
        m_stack[m_freq].pop();
        --m_freq_map[top_val];

        // update stack's max freq
        if (m_stack[m_freq].empty()) {
            --m_freq;
        }
        return top_val;
    }

private:
    size_t m_freq = 0;
    std::unordered_map<int, size_t> m_freq_map;
    std::unordered_map<int, std::stack<int>> m_stack;
};

/**
 * Your FreqStack object will be instantiated and called as such:
 * FreqStack* obj = new FreqStack();
 * obj->push(val);
 * int param_2 = obj->pop();
 */

File: LeetCode/Hard/0895-maximum-frequency-stack/0895-maximum-frequency-stack.py
"""
Motivation: Frequency Stacks.
      Use a hashmap to count each element's frequency and a hashmap of
      stacks where each key is a frequency level and the value is a stack
      of elements pushed at that frequency. Keep track of the max frequency.
      When pushing, increment the element's frequency and push it to the
      corresponding frequency stack. When poping, pop from the max frequency
      stack and decrement the element's frequency. If the max frequency stack
      becomes empty, decrement max frequency.
Complexity: 
    Time - O(1): All operations are constant time.
    Space - O(n): Storing each element in the frequency map and stacks.
"""
class FreqStack:

    def __init__(self):
        self.m_freq = 0
        self.m_freq_map = defaultdict(int)
        self.m_stack = defaultdict(list)
        
    def push(self, val: int) -> None:
        # get val's freq - map default value is 0
        self.m_freq_map[val] += 1
        val_freq = self.m_freq_map[val]

        # push to the right freq stack
        self.m_stack[val_freq].append(val)

        # update stack's max freq
        self.m_freq = max(self.m_freq, val_freq)
    
    def pop(self) -> int:
        # extract top value
        top_val = self.m_stack[self.m_freq].pop()
        self.m_freq_map[top_val] -= 1

        # update stack's max freq
        if (len(self.m_stack[self.m_freq]) == 0):
            self.m_freq -= 1

        return top_val
    
